installer: autorun xserver in graphical media

worldofpeace opened the pull request:
###### Motivation for this change
When I first used the NixOS iso I found it confusing and strange to have to do `systemctl start display-manager`. I assumed that I would have went straight to the graphical environment.

If anyone has a reason as to why it shouldn't do this please share.

###### Notes
In #42610 @adisbladis suggested:
> I think we should add another boot entry so you can still choose not to start the graphical environment, at the same time we make the one that autostarts the graphical environment the default.

Wouldn't this need https://github.com/NixOS/nixpkgs/pull/45345 to include that configuration as a child config in the grub menu?

Though I feel that might be to complex.


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---
cc @samueldr @xeji @grahamc

worldofpeace committed:
installer: autorun xserver in graphical media

samueldr commented:
I would like to know why it wasn't the default. There is probably a reason behind, and knowing it helps ensuring we don't cause any unintended issues.

* * *

Though, I think this requires one important sub-feature: the installer media should present an option that disables the `display-manager.service` so one can use the graphical iso without having X started; in case starting X is causing issues. IIRC this can be done through a [`systemd.mask=display-manager.service`](https://www.freedesktop.org/software/systemd/man/kernel-command-line.html) option on the kernel command line.

samueldr commented:
Yes, sorry, bad habit, jumped the gun and added my thoughts before reading that this was already in the body of your PR.

This is one way the boot option can be implemented.

worldofpeace commented:
No problem, I'm guilty of the same thing too when if I was busy.

worldofpeace commented:
If there's an agreement to needing this feature implicitly then I'd say that this is blocked until
the tests for https://github.com/NixOS/nixpkgs/pull/45345 are reviewed and it's merged.

samueldr commented:
Probably no relation. The installer iso has a different menu generation entirely from the installed system.

[Around here](https://github.com/NixOS/nixpkgs/blob/aec67d299c24d0bfff6f5ae04b5c462f3d3138d9/nixos/modules/installer/cd-dvd/iso-image.nix#L49-L65) a new option can be added (class is used to get an icon in the menu).

mmahut commented:
Any updates on this pull request, please?

worldofpeace commented:
I plan to pick this up again after the 19.09 release.

worldofpeace committed:
nixos/iso-image: add menu entry to disable display-manager

worldofpeace commented:
Ok, rebased and added a quirk to disable the display-manager.
I think we might've discussed on having a way to condition menu entries on whether it's for the graphical media or minimal. Though I still think this change won't do much harm without that feature.

worldofpeace commented:
I've tested running the graphical media
* Does it autorun X starting plasma5 :heavy_check_mark: 
* Did enabling the quirk send to me to the tty :heavy_check_mark:

worldofpeace committed:
rl-2003: graphical image autoruns X

samueldr commented on nixos/modules/installer/cd-dvd/iso-image.nix:
```diff
@@ -284,6 +284,14 @@ let
         ${buildMenuAdditionalParamsGrub2 config "video=1920x1080@60"}
       }
 
+      # If we boot into a graphical environment where X is autorunned
```
(it's not an issue per se) should it be autoran? or rephrased? English is weird, even weirderer when neologisms are introduced! ;)

worldofpeace replied:
I believe it's autoran, but it appears I've used multiple alternates even in the commit messages.

samueldr commented on nixos/modules/installer/cd-dvd/iso-image.nix:
```diff
@@ -284,6 +284,14 @@ let
+      # If we boot into a graphical environment where X is autorunned
+      # and always crashes, it makes the media unusable. Allow the user
+      # to disable this.
+      submenu "Disable display-manager" --class quirk-disable-displaymanager {
```
We can introduce an icon, if desired, and if one can be composed. They are composed from the [breeze icon theme](https://github.com/KDE/breeze-icons). If you see something in it that can be adapted, or used as an inspiration I can look into [adding it to the relevant artwork](https://github.com/NixOS/nixos-artwork/tree/master/bootloader/grub2-installer).

worldofpeace replied:
I found breeze/devices/64/video-display.svg the most suitable.
I mostly wanted something like just a blank monitor, or communicates a tty.

worldofpeace replied:
I'm highly partial to elementary's icons which appears to be exactly what I'm looking for

![Screenshot from 2019-09-28 21 14 48](https://user-images.githubusercontent.com/28888242/65824326-1e3d7900-e235-11e9-9c3f-8b7563553b88.png)

samueldr replied:
Oh, forgot to specify the symbolic icons. Though the one from breeze, once made into outlines, may work, I'll try to remember to look into it later. (The PR can be merged without that change.)

worldofpeace replied:
Ahh, there's computer/video-display symbolic icons that should be fine then.
I'll add an issue at nixos-artwork for that once merged @samueldr.

samueldr commented on nixos/doc/manual/release-notes/rl-2003.xml:
```diff
@@ -23,6 +23,14 @@
+   <listitem>
+    <para>
+      The graphical installer image starts the graphical session automatically.
+      Before you'd be greeted by a tty and asked to enter <command>systemectl start display-manager</command>.
```
```suggestion
      Before you'd be greeted by a tty and asked to enter <command>systemctl start display-manager</command>.
```

samueldr commented on nixos/doc/manual/release-notes/rl-2003.xml:
```diff
@@ -23,6 +23,14 @@
+    <para>
+      The graphical installer image starts the graphical session automatically.
+      Before you'd be greeted by a tty and asked to enter <command>systemectl start display-manager</command>.
+      It's still possible to disable the display-manager from running by selecting the <literal>Disable display-manager</literal>
```
```suggestion
      It is now possible to disable the display-manager from running by selecting the <literal>Disable display-manager</literal>
```

samueldr commented on nixos/doc/manual/release-notes/rl-2003.xml:
```diff
@@ -23,6 +23,14 @@
+      The graphical installer image starts the graphical session automatically.
+      Before you'd be greeted by a tty and asked to enter <command>systemectl start display-manager</command>.
+      It's still possible to disable the display-manager from running by selecting the <literal>Disable display-manager</literal>
+      quirk in GRUB.
```
```suggestion
      quirk in the boot menu.
```

GRUB is an implementation detail; though unlikely some users may not know what a GRUB is.

samueldr reviewed (changes requested):
Minor stuff, generally approving the change.

worldofpeace commented:
@samueldr My English and personality is literally a neologism :rofl: 

Thanks for the suggestions, will try to implement.

worldofpeace commented:
Rebased and resolved the conflicts in the release note.
I plan to integrate this some time today if there's no issues raised.
